src/main/scala/zio/actors/ActorRef.scala

```diff
@@ -0,0 +1,133 @@
+package zio.actors
+
+import java.io.{ IOException, ObjectInputStream, ObjectOutputStream, ObjectStreamException }
+
+import zio.nio.{ InetAddress, InetSocketAddress, SocketAddress }
+import zio.nio.channels.AsynchronousSocketChannel
+import zio.{ DefaultRuntime, IO, Task, UIO }
+
+/**
+ *
+ * Reference to actor that might reside on local JVM instance or be available via remote communication
+ *
+ * @tparam E error type
+ * @tparam F wrapper type constructing DSL
+ */
+sealed trait ActorRef[E <: Throwable, -F[+_]] extends Serializable {
+
+  /**
+   *
+   * Send a message to actor - it can perform `fire-and-forget` interaction pattern or `ask` pattern
+   *
+   * @param fa message
+   * @tparam A return type
+   * @return effectful response
+   */
+  def ![A](fa: F[A]): Task[A]
+
+  /**
+   * Get referential absolute actor path
+   * @return
+   */
+  def path: UIO[String]
+
+}
+
+/* INTERNAL API */
+
+private[actors] object ActorRefSerial {
+
+  private[actors] val runtimeForResolve = new DefaultRuntime {}
+
+}
+
+private[actors] sealed abstract class ActorRefSerial[E <: Throwable, -F[+_]](private var actorPath: String)
+    extends ActorRef[E, F]
+    with Serializable {
+
+  import ActorSystemUtils._
+
+  @throws[IOException]
+  protected def writeObject1(out: ObjectOutputStream): Unit =
+    out.writeObject(actorPath)
+
+  @throws[IOException]
+  protected def readObject1(in: ObjectInputStream): Unit = {
+    val rawActorPath = in.readObject()
+    actorPath = rawActorPath.asInstanceOf[String]
+  }
+
+  @throws[ObjectStreamException]
+  protected def readResolve1(): Object = {
+
+    val remoteRef = for {
+      resolved           <- resolvePath(actorPath)
+      (_, addr, port, _) = resolved
+      address <- InetAddress
+                  .byName(addr)
+                  .flatMap(iAddr => SocketAddress.inetSocketAddress(iAddr, port))
+    } yield ActorRefRemote[E, F](actorPath, address)
+
+    ActorRefSerial.runtimeForResolve.unsafeRun(remoteRef)
+
+  }
+
+  override def path: UIO[String] = UIO.effectTotal(actorPath)
+
+}
+
+private[actors] case class ActorRefLocal[E <: Throwable, -F[+_]](private val actorName: String, actor: Actor[E, F])
+    extends ActorRefSerial[E, F](actorName) {
+
+  override def ![A](fa: F[A]): Task[A] = actor ! fa
+
+  @throws[IOException]
+  private def writeObject(out: ObjectOutputStream): Unit =
+    super.writeObject1(out)
+
+  @throws[IOException]
+  private def readObject(in: ObjectInputStream): Unit =
+    super.readObject1(in)
+
+  @throws[ObjectStreamException]
+  private def readResolve(): Object =
+    super.readResolve1()
+
+}
+
+private[actors] case class ActorRefRemote[E <: Throwable, -F[+_]](
```

Comment:
So if the implementation is in ZIO-NIO, then we're not returning `Throwable` there (ever). It might happen, but it's doubtful you can do anything about it. So the signature looks a little suspicious and in my opinion should be at least upgraded to `Exception`.

Reply:
So right now while working with remote actors there are several error that can occur. I've written test cases for that in remote test suite, e.g. selecting remote actor from another `ActorSystem` that does not exsist or no actor on local `ActorSystem`. So they're basically errors related to remote functionality. 

Then also every actor has it's own error domain: `Actor[State, Errors, Messages]` which  can be some sealed trait. 

When sending a message to an actor the signature is: `def ![A](fa: F[A]): Task[A]` where `Task[A]` is `IO[Throwable, A]` because the error might be related to remoting issues or user defined errors. 

To join these two domains I decided to introduce `E` with upper type bound to `Throwable` so that user defined errors that occur in `Stateful` behavior descriptions fit to expected `Throwable` in `Task`.

The first test in remote test suite is using user defined errors: 
```scala
sealed trait MyErrorDomain extends Throwable
case object DomainError    extends MyErrorDomain
``` 

So when while sending a message an error occurs we can pattern match so see whether is user defined error or remote system related error.

And why `Throwable`? I decided to use it because `Task[A]` already uses it and will be more handy than `IO[Exception, A]`


So we end up with such tree:
```
                        Throwable
                    /                \
                   /                  \
   User defined                          Exception
    errors                           /               \
                                    /                 \ 
                         No such             No such remote
                         remote actor          actor sytem
```

I've been thinking about it but I couldn't come up with anything better.
Do you think about it?